File: src/module/account/account_protocol.erl
-module(account_protocol).
-export([read/2, write/2]).


-spec read(Protocol :: non_neg_integer(), Binary :: binary()) -> {ok, [integer() | binary() | list()]} | {error, Protocol :: non_neg_integer(), Binary :: binary()}.
read(10000, <<>>) ->
    {ok, []};

read(10001, <<ServerId:16, AccountNameLength:16, AccountName:AccountNameLength/binary>>) ->
    {ok, [ServerId, AccountName]};

read(10002, <<RoleNameLength:16, RoleName:RoleNameLength/binary, ServerId:16, AccountNameLength:16, AccountName:AccountNameLength/binary, Sex:8, Classes:8, ChannelLength:16, Channel:ChannelLength/binary, DeviceIdLength:16, DeviceId:DeviceIdLength/binary, MacLength:16, Mac:MacLength/binary, DeviceTypeLength:16, DeviceType:DeviceTypeLength/binary>>) ->
    {ok, [RoleName, ServerId, AccountName, Sex, Classes, Channel, DeviceId, Mac, DeviceType]};

read(10003, <<RoleId:64, RoleNameLength:16, RoleName:RoleNameLength/binary, ServerId:16, AccountNameLength:16, AccountName:AccountNameLength/binary>>) ->
    {ok, [RoleId, RoleName, ServerId, AccountName]};

read(10004, <<>>) ->
    {ok, []};

read(Protocol, Binary) ->
    {error, Protocol, Binary}.


-spec write(Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()) -> {ok, binary()} | {error, Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()}.
write(10000, Result) ->
    {ok, protocol:pack(10000, <<(protocol:text(Result))/binary>>)};

write(10001, [Result, List]) ->
    ListBinary = protocol:write_list(fun({RoleId, RoleName}) -> <<RoleId:64, (byte_size(RoleName)):16, (RoleName)/binary>> end, List),
    {ok, protocol:pack(10001, <<(protocol:text(Result))/binary, ListBinary/binary>>)};

write(10002, [Result, RoleId, RoleName]) ->
    {ok, protocol:pack(10002, <<(protocol:text(Result))/binary, RoleId:64, (byte_size(RoleName)):16, (RoleName)/binary>>)};

write(10003, Result) ->
    {ok, protocol:pack(10003, <<(protocol:text(Result))/binary>>)};

write(10004, Result) ->
    {ok, protocol:pack(10004, <<(protocol:text(Result))/binary>>)};

write(Protocol, Data) ->
    {error, Protocol, Data}.

File: script/make/protocol/json/AccountProtocol.js
let accountProtocol = {
    "read" : {
        "10000" : [],
        "10001" : [
            {"name" : "serverId", "type" : "u16", "comment" : "服务器ID", "explain" : []},
            {"name" : "accountName", "type" : "bst", "comment" : "账户名", "explain" : []}
        ],
        "10002" : [
            {"name" : "serverId", "type" : "u16", "comment" : "服务器ID", "explain" : []},
            {"name" : "sex", "type" : "u8", "comment" : "性别", "explain" : []},
            {"name" : "career", "type" : "u8", "comment" : "职业", "explain" : []},
            {"name" : "channelId", "type" : "u16", "comment" : "渠道ID", "explain" : []},
            {"name" : "name", "type" : "bst", "comment" : "名字", "explain" : []},
            {"name" : "nick", "type" : "bst", "comment" : "昵称", "explain" : []},
            {"name" : "deviceId", "type" : "bst", "comment" : "设备", "explain" : []},
            {"name" : "mac", "type" : "bst", "comment" : "mac地址", "explain" : []},
            {"name" : "deviceType", "type" : "bst", "comment" : "设备类型", "explain" : []}
        ]
    },
    "write" : {
        "10000" : [],
        "10001" : [
            {"name" : "result", "type" : "u8", "comment" : "结果(1成功/2:没有找到账户/3:登录禁止/4:重复登录)", "explain" : []}
        ],
        "10002" : [
            {"name" : "result", "type" : "u8", "comment" : "结果(1:成功/2:长度不对/3:字符问题/4:敏感词/5:名字重复)", "explain" : []}
        ]
    }
};